src/kg_topology_toolbox/topology_toolbox.py

```diff
@@ -5,108 +5,305 @@
 Topology toolbox main functionalities
 """
 
-from collections.abc import Iterable
+import warnings
+from functools import cache
 
 import numpy as np
 import pandas as pd
+from pandas.api.types import is_integer_dtype
 from scipy.sparse import coo_array
 
-from kg_topology_toolbox.utils import composition_count, jaccard_similarity
+from kg_topology_toolbox.utils import (
+    aggregate_by_relation,
+    composition_count,
+    jaccard_similarity,
+    node_degrees_and_rels,
+)
 
 
 class KGTopologyToolbox:
     """
-    Toolbox class to compute various Knowledge Graph topology statistics.
+    Toolbox class to compute Knowledge Graph topology statistics.
     """
 
-    def node_degree_summary(
-        self, df: pd.DataFrame, return_relation_list: bool = False
-    ) -> pd.DataFrame:
+    def __init__(
+        self,
+        kg_df: pd.DataFrame,
+        head_column: str = "h",
+        relation_column: str = "r",
+        tail_column: str = "t",
+    ):
+        """
+        Instantiate the Topology Toolbox for a Knowledge Graph defined
+        by the list of its edges (h,r,t).
+
+        :param kg_df:
+            A Knowledge Graph represented as a pd.DataFrame.
+            Must contain at least three columns, which specify the IDs of
+            head entity, relation type and tail entity for each edge.
+        :param head_column:
+            The name of the column with the IDs of head entities. Default: "h".
+        :param head_column:
+            The name of the column with the IDs of relation types. Default: "r".
+        :param head_column:
+            The name of the column with the IDs of tail entities. Default: "t".
+
+        """
+        for col_name in [head_column, relation_column, tail_column]:
+            if col_name in kg_df.columns:
+                if not is_integer_dtype(kg_df[col_name]):
+                    raise TypeError(
+                        f"Column {col_name} needs to be of an integer dtype"
+                    )
+            else:
+                raise ValueError(f"DataFrame {kg_df} has no column named {col_name}")
+        self.df = kg_df[[head_column, relation_column, tail_column]].rename(
+            columns={head_column: "h", relation_column: "r", tail_column: "t"}
+        )
+        if self.df.duplicated().any():
+            warnings.warn(
+                "The Knowledge Graph contains duplicated edges"
+                " -- some functionalities may produce incorrect results"
+            )
+        self.n_entity = self.df[["h", "t"]].max().max() + 1
+        self.n_rel = self.df.r.max() + 1
```

Comment:
Really like the error checking we are now doing but maybe this could also be moved into utils to make the class a bit cleaner.

Reply:
Thanks, it's a good idea - I refactored it as suggested, please let me know what you think now

Reply:
yup looks great thanks @AlCatt91!